### ISSUE TITLE
tf.while loop in MultiWorkerMirroredStrategy leads to training error: Complete shape not known for Adam/allreduce/CollectiveReduce

### ISSUE THREAD

**Issue by antonaldo** [tensorflow/tensorflow#48279]
**System information**

- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04.2 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.4.0-49-g85c8b2a817f 2.4.1
- Python version: 3.8.5
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: cuda 11.0/cudnn 8
- GPU model and memory: NVIDIA GTX 1080 8GB

**Describe the current behavior:**

In a cutom tf.keras.Model using a tf.while loop within train_step works fine when using non ditributed training or tf.distribute.MirroredStrategy.
When using tf.distribute.MultiWorkerMirroredStrategy following error is shown upon training startup which then leads to NaN loss when training a complex model:

> 2021-04-02 12:56:07.219137: W tensorflow/core/grappler/optimizers/scoped_allocator_optimizer.cc:455] error: Aborted: Complete shape not known for Adam/allreduce/CollectiveReduce
> 2021-04-02 12:56:07.219172: W tensorflow/core/grappler/optimizers/scoped_allocator_optimizer.cc:1138] error: Aborted: Complete shape not known for Adam/allreduce/CollectiveReduce
> 2021-04-02 12:56:07.219180: E tensorflow/core/grappler/optimizers/scoped_allocator_optimizer.cc:1155] ScopedAllocatorOptimizer: Aborted: Complete shape not known for Adam/allreduce/CollectiveReduce
> 2021-04-02 12:56:07.219187: W tensorflow/core/grappler/optimizers/scoped_allocator_optimizer.cc:928] error: Aborted: Complete shape not known for Adam/allreduce/CollectiveReduce

I have prepared an example to reproduce the problem in the code below. The issue is present for GPU and CPU. Due to the simplified example, it does not lead to immediate NaN loss as it does in my more complex model. However, I hope it is sufficient to reproduce and identify the issue. When switching to SGD the error message does not appear. Nevertheless, when using tf.distribute.MultiWorkerMirroredStrategy it goes to immediate NaN loss in my more complex model as well.

**Describe the expected behavior:**

Training should work with tf.distribute.MultiWorkerMirroredStrategy as it does for standalone training or training when using tf.distribute.MirroredStrategy.

**Standalone code to reproduce the issue**
Needed files are provided as well in this zipfile:
[example.zip](https://github.com/tensorflow/tensorflow/files/6249694/example.zip)
Use code below or unzip the files in a directory and run python test_worker_0.py on one instance and python test_worker_1.py on another instance of the same machine.

Code for TestModel.py:
```
import tensorflow as tf
import numpy as np

loss_tracker = tf.keras.metrics.Mean(name="loss")


class TestModel(tf.keras.Model):
    def __init__(self, *args, **kwargs):
        super(TestModel, self).__init__(*args, **kwargs)
        self.Dense = tf.keras.layers.Dense(80)
        self.MSE = tf.keras.losses.MeanSquaredError(reduction=tf.keras.losses.Reduction.NONE)

    def train_step(self, data):
        batch_size = tf.squeeze(tf.slice(tf.shape(data), [0], [1]), -1)
        #tried hardcoding batch_size to e.g. 16 with two workers -> does not resolve the issue
        max_length = 10
        i_start = tf.constant(0, dtype=tf.int32)
        dummy_inputs = tf.zeros([batch_size, 1, 200])
        gen_outputs = tf.zeros([batch_size, 0, 80], tf.float32)

        def body(i, input, output_full):
            output_single = self.Dense(input)
            output_full = tf.concat([output_full, output_single], 1)
            i_next = i + 1
            return i_next, input, output_full

        with tf.GradientTape() as tape:
            _, _, gen_data = tf.while_loop(cond=lambda i,
                                                       input,
                                                       output_full: tf.less(i, max_length),
                                           body=body,
                                           loop_vars=(i_start,
                                                      dummy_inputs,
                                                      gen_outputs,),
                                           shape_invariants=(i_start.get_shape(),
                                                             tf.TensorShape([None, None, 200]),
                                                             tf.TensorShape([None, None, 80])))
            loss = self.MSE(data, gen_data)
        grads = tape.gradient(loss, self.trainable_variables)
        self.optimizer.apply_gradients(zip(grads, self.trainable_variables))
        loss_tracker.update_state(loss)

        return {"loss": loss_tracker.result()}


def get_dummy_data():
    x = np.random.random((128, 10, 80))
    dataset = tf.data.Dataset.from_tensor_slices(x)
    dataset.options().experimental_distribute.auto_shard_policy = tf.data.experimental.AutoShardPolicy.DATA
    dataset = dataset.batch(32)
    return dataset


def get_model():
    model = TestModel()
    optimizer = tf.keras.optimizers.Adam()
    model.compile(optimizer=optimizer)
    return model

```

Code for test_worker_0.py:
```
import os
import json
import tensorflow as tf
os.environ.pop('TF_CONFIG', None)
os.environ['CUDA_VISIBLE_DEVICES'] = '-1'
tf_config = {
    'cluster': {
        'worker': ['localhost:12345', 'localhost:23456']
    },
    'task': {'type': 'worker', 'index': 0}
}
os.environ['TF_CONFIG'] = json.dumps(tf_config)

communication_options = tf.distribute.experimental.CommunicationOptions(
    implementation=tf.distribute.experimental.CommunicationImplementation.RING)
strategy = tf.distribute.MultiWorkerMirroredStrategy(communication_options=communication_options)

import TestModel

with strategy.scope():
    model = TestModel.get_model()

data = TestModel.get_dummy_data()
model.fit(data, epochs=4)
```

Code for test_worker_1.py:
```
import os
import json
import tensorflow as tf
os.environ.pop('TF_CONFIG', None)
os.environ['CUDA_VISIBLE_DEVICES'] = '-1'
tf_config = {
    'cluster': {
        'worker': ['localhost:12345', 'localhost:23456']
    },
    'task': {'type': 'worker', 'index': 1}
}
os.environ['TF_CONFIG'] = json.dumps(tf_config)

communication_options = tf.distribute.experimental.CommunicationOptions(
    implementation=tf.distribute.experimental.CommunicationImplementation.RING)
strategy = tf.distribute.MultiWorkerMirroredStrategy(communication_options=communication_options)

import TestModel

with strategy.scope():
    model = TestModel.get_model()

data = TestModel.get_dummy_data()
model.fit(data, epochs=4)
```

**Other info / logs**
Full logs from a run of the example:

> 2021-04-02 12:56:04.115797: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
> 2021-04-02 12:56:04.797812: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
> 2021-04-02 12:56:04.798297: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
> 2021-04-02 12:56:04.819525: E tensorflow/stream_executor/cuda/cuda_driver.cc:328] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
> 2021-04-02 12:56:04.819543: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: test
> 2021-04-02 12:56:04.819548: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: test
> 2021-04-02 12:56:04.819598: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:200] libcuda reported version is: 460.39.0
> 2021-04-02 12:56:04.819613: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:204] kernel reported version is: 460.39.0
> 2021-04-02 12:56:04.819618: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:310] kernel version seems to match DSO: 460.39.0
> 2021-04-02 12:56:04.819910: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
> To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
> 2021-04-02 12:56:04.821691: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
> 2021-04-02 12:56:04.823450: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
> 2021-04-02 12:56:04.827242: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:301] Initialize GrpcChannelCache for job worker -> {0 -> localhost:12345, 1 -> localhost:23456}
> 2021-04-02 12:56:04.828921: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:411] Started server with target: grpc://localhost:23456
> 2021-04-02 12:56:04.885062: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
> 2021-04-02 12:56:04.885517: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 3700075000 Hz
> Epoch 1/4
> 2021-04-02 12:56:07.219137: W tensorflow/core/grappler/optimizers/scoped_allocator_optimizer.cc:455] error: Aborted: Complete shape not known for Adam/allreduce/CollectiveReduce
> 2021-04-02 12:56:07.219172: W tensorflow/core/grappler/optimizers/scoped_allocator_optimizer.cc:1138] error: Aborted: Complete shape not known for Adam/allreduce/CollectiveReduce
> 2021-04-02 12:56:07.219180: E tensorflow/core/grappler/optimizers/scoped_allocator_optimizer.cc:1155] ScopedAllocatorOptimizer: Aborted: Complete shape not known for Adam/allreduce/CollectiveReduce
> 2021-04-02 12:56:07.219187: W tensorflow/core/grappler/optimizers/scoped_allocator_optimizer.cc:928] error: Aborted: Complete shape not known for Adam/allreduce/CollectiveReduce
> 4/4 [==============================] - 2s 3ms/step - loss: 0.3312
> Epoch 2/4
> 4/4 [==============================] - 0s 3ms/step - loss: 0.3290
> Epoch 3/4
> 4/4 [==============================] - 0s 3ms/step - loss: 0.3270
> Epoch 4/4
> 4/4 [==============================] - 0s 3ms/step - loss: 0.3250

**Comment by amahendrakar** [comment]
> Due to the simplified example, it does not lead to immediate NaN loss as it does in my more complex model. However, I hope it is sufficient to reproduce and identify the issue.

@antonaldo,
I did not face the NaN loss issue on running the code for 100 epochs. Please check the attached screenshot for reference. 
![Screenshot 2021-04-06 12 26 26 AM](https://user-images.githubusercontent.com/57165142/113613608-17d05e80-966f-11eb-8545-0dd039f0f9ae.png)

**Comment by amahendrakar** [comment]
> 2021-04-02 12:56:04.819525: E tensorflow/stream_executor/cuda/cuda_driver.cc:328] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected

Also, looks like TensorFlow failed to detect the GPU on your machine. Could you please share the output of the below code with us?

```
import tensorflow as tf
print("Num GPUs Available: ", len(tf.config.list_physical_devices('GPU')))
```

Thanks!

**Comment by antonaldo** [comment]
@amahendrakar 

Thanks for looking into this. Please let me clarify the example and the issue.

Idea of the example was to reproduce this error message in a simple way:

> 2021-04-02 12:56:07.219180: E tensorflow/core/grappler/optimizers/scoped_allocator_optimizer.cc:1155] ScopedAllocatorOptimizer: Aborted: Complete shape not known for Adam/allreduce/CollectiveReduce

This example is related to the allreduce error observed. It is a very simplified example, so I am not sure if it will go into NaN-loss.
Can you reproduce the error message related to allreduce with the example?
GPU was turned off on purpose in my example, so that it becomes clear that this error message is not necessarily GPU. However, the allreduce error message shows op on GPU as well. This happens if I use the example one one machine or distributed on two different machines with GPU or CPU.
This only happens when using tf.distribute.MultiWorkerMirroredStrategy.
When using tf.distribute.MirroredStrategy or non ditributed training the error does not show up. So there seems to be an issue with tf.distribute.MultiWorkerMirroredStrategy when used in the context of the example provided.

I hope this clarifies the problem.

**Comment by nikitamaia** [comment]
Hi @antonaldo, your model trains fine with `MirroredStrategy` but then when you use `MultiWorkerMirroredStrategy` it still trains, but you see the message `ScopedAllocatorOptimizer: Aborted: Complete shape not known for Adam/allreduce/CollectiveReduce`. Have I understood this correctly?

**Comment by antonaldo** [comment]
@nikitamaia
Yes, you are correct, when using `MultiWorkerMirroredStrategy` the model performs the training loop, but the error message is displayed at the beginning. However, when using my full model with `MultiWorkerMirroredStrategy` loss is not decreasing, but ends up in NaN loss after a few steps.
When using `MirroredStrategy` training works as expected with my full model (loss decreases).
I assume that the observed error message is the reason for this unexpected behavior when `MultiWorkerMirroredStrategy` is used.

**Comment by nikitamaia** [comment]
So that message means that an optimization failed and your model training might run slower. I don't know that it would cause a nan loss though. A few questions:
- Do you still have the nan loss if you use a different optimizer?
- Can you also provide a screenshot of the logs for your full example?
- are you scaling your batch size by the number of replicas? `global_batch_size = per_replica_batch_size * strategy.num_replicas_in_sync`

**Comment by nikitamaia** [comment]
Quick update, looks like you should not encounter this ScopedAllocator issue with TF 2.5.
Can you try your code with 2.5 or nightly?

**Comment by antonaldo** [comment]
@nikitamaia:
I have tried it with nightly and the error is no longer displayed. Therefore, this problem is resolved an I will close this issue.
I realized that the observed NaN-loss in my full model is most likely related to use of `tf.keras.layers.experimental.SyncBatchNormalization()`. I will look into this and file a new issue if needed.
Thanks for your support!
